fn/async.go

```diff
@@ -0,0 +1,110 @@
+package fn
+
+import (
+	"context"
+	"errors"
+)
+
+// Async represents an action whose results are being computed outside this
+// control thread.
+type Async[A any] interface {
+	// Await may be called to block on the completion of the remote
+	// computation.
+	Await() (*A, error)
+
+	// Cancel may be called to abandon the remote computation.
+	Cancel()
+}
+
+var (
+	// ContextCancelled indicates that this Async was cancelled because the
+	// surrounding context was being torn down.
+	ContextCancelled = errors.New("context cancelled")
+)
+
+// RunAsyncWithCancel is a function that will transform a synchronous function
+// and make it a cancellable asynchronous call. This can be useful if the
+// function in the argument can block indefinitely and the calling context needs
+// a way to abandon waiting for it to complete.
+func RunAsyncWithCancel[A any](ctx context.Context,
+	f func(ctx context.Context) (*A, error)) Async[A] {
+
+	localCtx, cancel := context.WithCancel(ctx)
+
+	a := async[A]{
+		ctx:    localCtx,
+		cancel: cancel,
+		res:    make(chan *A, 1),
+		err:    make(chan error, 1),
+	}
+
+	go func() {
+		res, err := f(localCtx)
```

Comment:
so the provided function needs to cleanly listen to the Done() channel to actually be abortable ?

Reply:
If you do not listen on `context.Done`, then the goroutine will hang on whatever other channel signal it is waiting on to continue. The control flow of the Async itself will remain cancellable via the `Cancel` function producing an error for the thread that is hanging on `Await`, but a cancellation of the async action won't necessarily abort the inner thread unless `f` listens on `context.Done`

Reply:
That's the best we can do I guess, and the caller needs to be responsible as well.